doc/release/upcoming_changes/16938.c_api.rst

```diff
@@ -0,0 +1,17 @@
+Size of ``np.ndarray`` and ``np.void_`` changed
+-----------------------------------------------
+The size of the ``PyArrayObject`` and ``PyVoidScalarObject``
+structures have changed.  This also modifies the header
+definition::
+
+    #define NPY_SIZEOF_PYARRAYOBJECT (PyArray_Type.tp_basicsize)
+
+since the size should not be considered a compile time constant.
```

Comment:
```suggestion
Making this a compile-time constant may lead to bugs, but it is part of the public headers.
```

Reply:
How about:

> This was previously a compile-time constant, which could lead to bugs when running in a different NumPy version.

One thing we could do is adding a new `#define NPY_SIZEOF_PYARRAYOBJECT_STATIC sizeof(PyArrayObject_fields)` so that it is easier to replace the safe version with the unsafe one where necessary (it would be the users job to check this on import time).  But as I don't know anyone actually using it, I am somewhat OK with forcing users into `sizeof(PyArrayObject_fields)` using the deprecated API.

Reply:
I think any attempt at using this is doomed to failure. Can we deprecate the macro, with a message that it is the wrong thing to checke?

Reply:
I think we must disallow its use. These are exactly the use cases we want to loudly fail with the new code since they are depending on something staying the same size when it doesn't.

Reply:
@mattip I agree.  Did you formed a clear opinion on what to do here?  No `NPY_SIZEOF_PYARRAYOBJECT_STATIC` sounds good to me.  But I need to make a call what to do specifically for this PR.

* Additionally to making it a run-time constant (if that even matters for anyone), we add a deprecated tag in C?
* We just remove it (i.e. make it an `# error`)... The `(PyArray_Type.tp_basicsize)` doesn't have much point probably?
* Only deprecate it, which makes it easier to compile code that is not future proof without realizing it.

I am OK with both of the first versions, the `# error` might be nice, since it allows us to write a few words about what is going on in a very clear way that can't get lost.

Reply:
we could try to do some [static assert](https://stackoverflow.com/questions/3385515/static-assert-in-c) complicated thing, or just remove it and document the removal in the release notes. I think removing it is the easiest and will get the desired response if anyone is actually using it.

Reply:
OK, the last commit just deletes this `#define` and adds some info to the `PyArrayObject` structure documentation.

Not sure where we stand with putting this in (or more when), seems like not for 1.20 most likely, so will have to update the text slightly again before merging finally.